# Dataquest.io Guided Project 02

In [1]:
# Import the reader function from the csv module
from csv import reader

# Open and read hacker news file, hn
opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)

# Print first 5 rows
print(hn[0:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
# Remove headers and save
headers = hn[0]
hn = hn[1:]
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [3]:
# Print first five rows of hn
print(hn[0:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [4]:
# Create ask, show, and other lists
ask_posts = []
show_posts = []
other_posts = []

# Loop through hn
for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

# Check lengths of the three lists
print("Number of ask posts: ", len(ask_posts))
print("Number of show posts: ", len(show_posts))
print("Number of other posts: ", len(other_posts))

Number of ask posts:  1744
Number of show posts:  1162
Number of other posts:  17194


### Determine if the ask posts or show posts receive more comments on average

In [5]:
total_ask_comments = 0

# Loop over ask_posts and add up the numbers of comments
for row in ask_posts:
    num = int(row[4])
    total_ask_comments += num

# Compute average
avg_ask_comments = total_ask_comments / len(ask_posts)
print("The average number of comments on ask posts: ", avg_ask_comments)

The average number of comments on ask posts:  14.038417431192661


In [6]:
total_show_comments = 0

# Loop over ask_posts and add up the numbers of comments
for row in show_posts:
    num = int(row[4])
    total_show_comments += num

# Compute average
avg_show_comments = total_show_comments / len(show_posts)
print("The average number of comments on show posts: ", avg_show_comments)

The average number of comments on show posts:  10.31669535283993


### From the code above, one can see that the average number of comments received on ask posts (14.04) is greater than that received on show posts (10.32).

In [7]:
# Import datetime module
import datetime as dt

# result_list contains the time a comment is created and the number of posts
result_list = []

# Iterate over ask_posts, append time created and number to result_list
for row in ask_posts:
    time = row[6]
    num = row[4]
    result_list.append([time, num])
    
# Instantiate empty dictionaries
counts_by_hour = {}
comments_by_hour = {}

# Iterate over result_list
for row in result_list:
    # 10/15/2015 16:38
    dt_str = row[0]
    dt_obj = dt.datetime.strptime(dt_str, "%m/%d/%Y %H:%M")
    hour = dt_obj.strftime("%H")
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += int(row[1])
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = int(row[1])

In [47]:
# Compute average number of comments per post for posts created during each hour of the day
avg_by_hour = []

# Iterate over over counts
for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])

print(avg_by_hour)

[['01', 11.383333333333333], ['18', 13.20183486238532], ['17', 11.46], ['21', 16.009174311926607], ['03', 7.796296296296297], ['06', 9.022727272727273], ['07', 7.852941176470588], ['13', 14.741176470588234], ['20', 21.525], ['11', 11.051724137931034], ['04', 7.170212765957447], ['10', 13.440677966101696], ['16', 16.796296296296298], ['00', 8.127272727272727], ['08', 10.25], ['22', 6.746478873239437], ['09', 5.5777777777777775], ['23', 7.985294117647059], ['15', 38.5948275862069], ['19', 10.8], ['05', 10.08695652173913], ['12', 9.41095890410959], ['14', 13.233644859813085], ['02', 23.810344827586206]]


In [48]:
swap_avg_by_hour = []

# Loop over avg_by_hour, swap columns
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

# Print the swapped list
print(swap_avg_by_hour)

[[11.383333333333333, '01'], [13.20183486238532, '18'], [11.46, '17'], [16.009174311926607, '21'], [7.796296296296297, '03'], [9.022727272727273, '06'], [7.852941176470588, '07'], [14.741176470588234, '13'], [21.525, '20'], [11.051724137931034, '11'], [7.170212765957447, '04'], [13.440677966101696, '10'], [16.796296296296298, '16'], [8.127272727272727, '00'], [10.25, '08'], [6.746478873239437, '22'], [5.5777777777777775, '09'], [7.985294117647059, '23'], [38.5948275862069, '15'], [10.8, '19'], [10.08695652173913, '05'], [9.41095890410959, '12'], [13.233644859813085, '14'], [23.810344827586206, '02']]


In [49]:
# Sort swapped list in reverse order
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

# Import datetime module
import datetime as dt

print("Top 5 Hours for Ask Posts Comments")
for l in sorted_swap[0:5]:
    time_dt = dt.datetime.strptime(l[1], "%H")
    time_dt_1 = time_dt.time()
    time = dt.time.strftime(time_dt_1, "%H:%m")
    comments = int(l[0])
    print('{1}: {0:.2f} average comments per post'.format(comments, time))

Top 5 Hours for Ask Posts Comments
15:01: 38.00 average comments per post
02:01: 23.00 average comments per post
20:01: 21.00 average comments per post
16:01: 16.00 average comments per post
21:01: 16.00 average comments per post
